sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/MarathonConstraintParser.java

```diff
@@ -0,0 +1,321 @@
+    }
+
+    /**
+     * Interface for generating a PlacementRule for a given marathon operator such as UNIQUE or CLUSTER.
+     */
+    private interface Operator {
+        public PlacementRule run(String fieldName, String operatorName, Optional<String> parameter) throws IOException;
+    }
+
+    /**
+     * {@code UNIQUE} tells Marathon to enforce uniqueness of the attribute across all of an app's
+     * tasks. For example the following constraint ensures that there is only one app task running
+     * on each host: {@code [["hostname", "UNIQUE"]]}
+     */
+    private static class UniqueOperator implements Operator {
+        public PlacementRule run(String fieldName, String operatorName, Optional<String> ignoredParameter) {
+            if (isHostname(fieldName)) {
+                return new MaxPerHostnameRule(1);
+            } else {
+                // Ensure that:
+                // - Task sticks to nodes with matching fieldName defined at all (AttributeRule)
+                // - Task doesn't exceed one instance on those nodes (MaxPerAttributeRule)
+                StringMatcher matcher = RegexMatcher.createAttribute(fieldName, ".*");
+                return new AndRule(
+                        AttributeRule.require(matcher), new MaxPerAttributeRule(1, matcher));
+            }
+        }
+    }
+
+    /**
+     * {@code CLUSTER} allows you to run all of your app's tasks on agent nodes that share a certain
+     * attribute. This is useful for example if you have apps with special hardware needs, or if you
+     * want to run them on the same rack for low latency: {@code [["rack_id", "CLUSTER", "rack-1"]]}
+     *
+     * You can also use this attribute to tie an application to a specific node by using the
+     * hostname property: {@code [["hostname", "CLUSTER", "a.specific.node.com"]]}
+     */
+    private static class ClusterOperator implements Operator {
+        public PlacementRule run(String fieldName, String operatorName, Optional<String> requiredParameter)
+                throws IOException {
+            String parameter = validateRequiredParameter(operatorName, requiredParameter);
+            if (isHostname(fieldName)) {
+                return HostnameRule.require(ExactMatcher.create(parameter));
+            } else {
+                return AttributeRule.require(ExactMatcher.createAttribute(fieldName, parameter));
+            }
+        }
+    }
+
+    /**
+     * {@code GROUP_BY} can be used to distribute tasks evenly across racks or datacenters for high
+     * availability: {@code [["rack_id", "GROUP_BY"]]}
+     *
+     * Marathon only knows about different values of the attribute (e.g. "rack_id") by analyzing
+     * incoming offers from Mesos. If tasks are not already spread across all possible values,
+     * specify the number of values in constraints. If you don't specify the number of values, you
+     * might find that the tasks are only distributed in one value, even though you are using the
```

Comment:
Might be worth noting which circumstances could lead to this. I assume it's a function of the distribution of $ATTRIBUTE among the incoming offers?